pkg/sbom/cyclonedx/cyclonedx.go

```diff
@@ -0,0 +1,412 @@
+package cyclonedx
+
+import (
+	"sort"
+	"strconv"
+	"strings"
+
+	cdx "github.com/CycloneDX/cyclonedx-go"
+	"github.com/package-url/packageurl-go"
+	"golang.org/x/xerrors"
+
+	ftypes "github.com/aquasecurity/fanal/types"
+	dtypes "github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/purl"
+	"github.com/aquasecurity/trivy/pkg/sbom"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+const (
+	Namespace = "aquasecurity:trivy:"
+
+	PropertySchemaVersion = "SchemaVersion"
+	PropertyType          = "Type"
+	PropertyClass         = "Class"
+
+	// Image properties
+	PropertySize       = "Size"
+	PropertyImageID    = "ImageID"
+	PropertyRepoDigest = "RepoDigest"
+	PropertyDiffID     = "DiffID"
+	PropertyRepoTag    = "RepoTag"
+
+	// Package properties
+	PropertySrcName         = "SrcName"
+	PropertySrcVersion      = "SrcVersion"
+	PropertySrcRelease      = "SrcRelease"
+	PropertySrcEpoch        = "SrcEpoch"
+	PropertyModularitylabel = "Modularitylabel"
+	PropertyFilePath        = "FilePath"
+	PropertyLayerDigest     = "LayerDigest"
+	PropertyLayerDiffID     = "LayerDiffID"
+)
+
+const (
+	// https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times
+	timeLayout = "2006-01-02T15:04:05+00:00"
+
+	FormatCycloneDX sbom.SBOMFormat = "cyclonedx"
+)
+
+type TrivyBOM struct {
+	cdx.BOM
+}
+
+func (b TrivyBOM) parseComponents() (string, *ftypes.OS, map[string]*ftypes.Application, map[string]cdx.Component, error) {
```

Comment:
I feel like we should parse components after parsing dependencies. Actually, this method keeps components as is and defers parsing them. Then, similar logic are located in some places, making it harder to read the source code.